supabase/migrations/24_billing_historicals.sql

```diff
@@ -0,0 +1,230 @@
+
+-- Always use a transaction, y'all.
+begin;
+
+-- Historical record of tenant billing statements. This structure
+-- comes from the return value of `billing_report_202308`.
+create table billing_historicals (
+    billed_prefix       catalog_tenant  not null,
+    billed_month        timestamptz     not null,
+    report              jsonb           not null,
+
+    check (date_trunc('month', billed_month) = billed_month),
+    unique (billed_prefix, billed_month)
+);
+alter table billing_historicals enable row level security;
+grant all on billing_historicals to postgres;
+
+create policy "Users must be authorized to their catalog tenant"
+  on billing_historicals as permissive for select
+  using (exists(
+    select 1 from auth_roles('admin') r where billed_prefix ^@ r.role_prefix
+  ));
+grant select on billing_historicals to authenticated;
+
+-- Calculate the specified month's billing report for every tenant
+-- and save those reports to billing_historicals.
+create function internal.freeze_billing_month(billed_month timestamptz)
+returns integer as $$
+declare
+    tenant_row record;
+    tenant_count integer = 0;
+begin
+    for tenant_row in select tenant as tenant_name from tenants loop
+        tenant_count = tenant_count + 1;
+        insert into billing_historicals
+        select
+            tenant_row.tenant_name as billed_prefix,
```

Comment:
nit: can this be `report->>billed_prefix, report->>billed_month` or does that not work for some reason?